### PR TITLE
Allow to union `MultiIndex` with empty `RangeIndex`

### LINKED ISSUE #41234: Regression: concatenating `MultiIndex` with empty `RangeIndex` raises
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

As also reported here: https://github.com/dask/dask/issues/7610, PR https://github.com/pandas-dev/pandas/pull/38671/  introduced the following behaviour:

> A reproducer with just pandas:
> 
> ```
> df1 = pd.DataFrame([[1, 2]], columns=pd.MultiIndex.from_tuples([('B', 1), ('C', 1)]))
> df2 = pd.DataFrame(index=[0], columns=pd.RangeIndex(0))
> 
> pd.concat([df1, df2])
> ```
> 
> What triggers the error here is the `columns=pd.RangeIndex(0)` for the empty `df2` (by default pandas creates a zero-length object dtype Index, which works fine, but if it's an empty RangeIndex, we now get this error). 
> 
> This is a regression in itself in pandas. But I am also wondering a bit where the empty RangeIndex is coming from (it seems that the groupby operation in dask results in some empty partitions)

_Originally posted by @jorisvandenbossche in https://github.com/dask/dask/issues/7610#issuecomment-827708470_

IIUC the fix for this could possibly be as simple as replacing
https://github.com/pandas-dev/pandas/blob/0acbff881c9273c2cda962fb263f2862a5e7be28/pandas/core/indexes/base.py#L2926-L2928
with
```python
            if isinstance(self, ABCMultiIndex) and not is_object_dtype(
                unpack_nested_dtype(other)
            ) and len(other) > 0:
```

### PR DESCRIPTION
**Pull request by mlondschien**
- [x] closes #41234
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

See also https://github.com/dask/dask/issues/7610 and https://github.com/JDASoftwareGroup/kartothek/issues/464. The regression seems to have been introduced in  pandas-dev/pandas#38671.

### CODE REVIEW

File: pandas/tests/indexes/multi/test_setops.py [hunk 1]
```diff
@@ -414,6 +414,14 @@ def test_union_empty_self_different_names():
     tm.assert_index_equal(result, expected)
 
 
+def test_union_multiindex_empty_rangeindex():
+    # GH#41234
+    mi = MultiIndex.from_arrays([[1, 2], [3, 4]], names=["a", "b"])
```
Comment (jreback) [inline pandas/tests/indexes/multi/test_setops.py:419]: test the reverse also

File: doc/source/whatsnew/v1.3.0.rst [hunk 2]
```diff
@@ -754,6 +754,7 @@ Indexing
 - Bug in :meth:`DataFrame.loc.__setitem__` when setting-with-expansion incorrectly raising when the index in the expanding axis contains duplicates (:issue:`40096`)
 - Bug in :meth:`DataFrame.loc` incorrectly matching non-boolean index elements (:issue:`20432`)
 - Bug in :meth:`Series.__delitem__` with ``ExtensionDtype`` incorrectly casting to ``ndarray`` (:issue:`40386`)
+- Bug in :meth:`MultiIndex.union` raising for an empty ``RangeIndex`` (:issue:`41234`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:757]: is this a bug vc 1.2.x? if not then no whatsnew is needed (instead add this issue number next to the other release note)

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
also i see that the actual issue is showing with concat; can u add a test for that as well

File: pandas/tests/indexes/multi/test_setops.py [hunk 3]
```diff
@@ -414,6 +414,14 @@ def test_union_empty_self_different_names():
...
+    ri = pd.RangeIndex(0)
+    tm.assert_index_equal(mi, mi.union(ri), check_names=False)
```
Comment (jreback) [inline pandas/tests/indexes/multi/test_setops.py:421]: this is wrong, it should be of type Index or we should raise, this shouldn't just silently work

cc @jbrockmendel

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/multi/test_setops.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/indexes/multi/test_setops.py:421]: hmm we actually return the non-empty even when these are no the same type

Comment (mlondschien) [reply pandas/tests/indexes/multi/test_setops.py:421]: Not quite sure what you mean with

> it should be of type Index or we should raise

This should copy the behavior to 1.2.4.

Comment (jbrockmendel) [inline pandas/tests/indexes/multi/test_setops.py:421]: can you define `result = mi.union(ri)` on the previous line

Comment (jbrockmendel) [reply pandas/tests/indexes/multi/test_setops.py:421]: i think this is fine.  its values-dependent which we're trying to move away from, but it is consistent with what we do in all the other Index set methods

Comment (jreback) [reply pandas/tests/indexes/multi/test_setops.py]: yep

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mlondschien

**Pull request merged**